In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [10]:
train = pd.read_csv("https://raw.githubusercontent.com/Jaimin09/Coding-Lane-Assets/main/Multiple%20Linear%20Regression%20in%20Python/train_data.csv")
train.head()

Unnamed: 0  Id  MSSubClass  MSZoning  LotArea  Street  LotShape  \
0           0   1          60         3     8450       1         3   
1           1   2          20         3     9600       1         3   
2           2   3          60         3    11250       1         0   
3           3   4          70         3     9550       1         0   
4           4   5          60         3    14260       1         0   

   LandContour  Utilities  LotConfig  ...  EnclosedPorch  3SsnPorch  \
0            3          0          4  ...              0          0   
1            3          0          2  ...              0          0   
2            3          0          4  ...              0          0   
3            3          0          0  ...            272          0   
4            3          0          2  ...              0          0   

   ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  SaleType  SaleCondition  \
0            0         0        0       2    2008         8              4   
1            0         0        0       5    2007         8              4   
2            0         0        0       9    2008         8              4   
3            0         0        0       2    2006         8              0   
4            0         0        0      12    2008         8              4   

   SalePrice  
0  12.247694  
1  12.109011  
2  12.317167  
3  11.849398  
4  12.429216  

[5 rows x 72 columns]

In [11]:
# save to DataFrame to csv file
train.to_csv("train_dataset.csv",index=False)

In [8]:
test = pd.read_csv("https://raw.githubusercontent.com/Jaimin09/Coding-Lane-Assets/main/Multiple%20Linear%20Regression%20in%20Python/test_data.csv")
test.head()

Unnamed: 0    Id  MSSubClass  MSZoning  LotArea  Street  LotShape  \
0        1200  1202          60         3    10400       1         3   
1        1201  1203          50         4     6000       1         3   
2        1202  1204          20         3     9750       1         3   
3        1203  1205          20         3    10140       1         3   
4        1204  1206          20         3    14684       1         0   

   LandContour  Utilities  LotConfig  ...  EnclosedPorch  3SsnPorch  \
0            3          0          0  ...              0          0   
1            3          0          0  ...            208          0   
2            3          0          4  ...              0          0   
3            3          0          4  ...              0          0   
4            3          0          1  ...              0          0   

   ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  SaleType  SaleCondition  \
0            0         0        0       3    2009         8              4   
1            0         0        0       5    2009         8              4   
2            0         0        0      10    2009         8              4   
3            0         0        0       7    2006         8              4   
4            0         0        0       6    2009         8              4   

   SalePrice  
0  12.195517  
1  11.669929  
2  12.269047  
3  11.941456  
4  12.513190  

[5 rows x 72 columns]

In [9]:
# save dataframe to csv file in current directory
test.to_csv("test_dataset.csv",index=False)

In [13]:
# delete the unwanted feature from train and test data
train = train.drop(["Unnamed: 0", "Id"], axis=1)
test = test.drop(["Unnamed: 0", "Id"],axis= 1)

In [15]:
train.head()


MSSubClass  MSZoning  LotArea  Street  LotShape  LandContour  Utilities  \
0          60         3     8450       1         3            3          0   
1          20         3     9600       1         3            3          0   
2          60         3    11250       1         0            3          0   
3          70         3     9550       1         0            3          0   
4          60         3    14260       1         0            3          0   

   LotConfig  LandSlope  Neighborhood  ...  EnclosedPorch  3SsnPorch  \
0          4          0             5  ...              0          0   
1          2          0            24  ...              0          0   
2          4          0             5  ...              0          0   
3          0          0             6  ...            272          0   
4          2          0            15  ...              0          0   

   ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  SaleType  SaleCondition  \
0            0         0        0       2    2008         8              4   
1            0         0        0       5    2007         8              4   
2            0         0        0       9    2008         8              4   
3            0         0        0       2    2006         8              0   
4            0         0        0      12    2008         8              4   

   SalePrice  
0  12.247694  
1  12.109011  
2  12.317167  
3  11.849398  
4  12.429216  

[5 rows x 70 columns]

In [16]:
test.head()

MSSubClass  MSZoning  LotArea  Street  LotShape  LandContour  Utilities  \
0          60         3    10400       1         3            3          0   
1          50         4     6000       1         3            3          0   
2          20         3     9750       1         3            3          0   
3          20         3    10140       1         3            3          0   
4          20         3    14684       1         0            3          0   

   LotConfig  LandSlope  Neighborhood  ...  EnclosedPorch  3SsnPorch  \
0          0          0             5  ...              0          0   
1          0          0             3  ...            208          0   
2          4          0             5  ...              0          0   
3          4          0            14  ...              0          0   
4          1          0            20  ...              0          0   

   ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  SaleType  SaleCondition  \
0            0         0        0       3    2009         8              4   
1            0         0        0       5    2009         8              4   
2            0         0        0      10    2009         8              4   
3            0         0        0       7    2006         8              4   
4            0         0        0       6    2009         8              4   

   SalePrice  
0  12.195517  
1  11.669929  
2  12.269047  
3  11.941456  
4  12.513190  

[5 rows x 70 columns]

In [17]:
# And now let's separate the X and Y.
train_data = train.values
Y = train_data[:,-1].reshape(train_data.shape[0],1)
X = train_data[:,:-1]

In [18]:
test_data = test.values
Y_test = test_data[:, -1].reshape(test_data.shape[0], 1)
X_test = test_data[:, :-1]

In [19]:
print("Shape of X_train :", X.shape)
print("Shape of Y_train :", Y.shape)
print("Shape of X_test :", X_test.shape)
print("Shape of Y_test :", Y_test.shape)

Shape of X_train : (1200, 69)
Shape of Y_train : (1200, 1)
Shape of X_test : (258, 69)
Shape of Y_test : (258, 1)


# So, we have data of 1200 houses in our dataset, and each house has 69
# features. Similarly, we have 258 houses in our dataset.

In [20]:
X = np.vstack((np.ones((X.shape[0], )), X.T)).T
X_test = np.vstack((np.ones((X_test.shape[0], )), X_test.T)).T

In [21]:
import numpy as np

def model(X, Y, learning_rate, iteration):
    m = Y.size  # Number of training examples
    theta = np.zeros((X.shape[1], 1))  # Initialize model parameters (weights)
    cost_list = []  # List to store the cost at each iteration

    for i in range(iteration):
        y_pred = np.dot(X, theta)  # Predict the output using current parameters
        cost = (1/(2*m)) * np.sum(np.square(y_pred - Y))  # Compute the cost (mean squared error)
        d_theta = (1/m) * np.dot(X.T, y_pred - Y)  # Compute the gradient
        theta = theta - learning_rate * d_theta  # Update the parameters using the gradient and learning rate
        cost_list.append(cost)  # Append the current cost to the list

        # Print the cost for every 10% of the iterations
        if i % (iteration // 10) == 0:
            print("Cost is :", cost)

    return theta, cost_list  # Return the final parameters and the cost list

# Example usage:
# X = ... (Your feature matrix)
# Y = ... (Your target variable)
# learning_rate = 0.01
# iteration = 1000
# theta, cost_list = model(X, Y, learning_rate, iteration)


In [22]:
iteration = 10000
learning_rate = 0.000000005
theta, cost_list = model(X, Y, learning_rate = learning_rate, iteration =
iteration)

Cost is : 72.37539364066856
Cost is : 0.02790416831031685
Cost is : 0.01725106537214416
Cost is : 0.01635527270554828
Cost is : 0.016158836087530753
Cost is : 0.01604095849845061
Cost is : 0.01594682732375345
Cost is : 0.015867896317230023
Cost is : 0.01580056801478539
Cost is : 0.015742355306482908
